chain/client/src/stateless_validation/endorsement_tracker.rs

```diff
@@ -0,0 +1,141 @@
+// Ideally, we should not be processing more than num_shards chunks at a time.
+const NUM_CHUNK_ENDORSEMENTS_CACHE_COUNT: usize = 100;
+
+pub struct EndorsementTracker {
+    epoch_manager: Arc<dyn EpochManagerAdapter>,
+    /// We store the validated chunk endorsements received from chunk validators
+    /// This is keyed on chunk_hash and account_id of validator to avoid duplicates.
+    /// Chunk endorsements would later be used as a part of block production.
+    chunk_endorsements: lru::LruCache<ChunkHash, HashMap<AccountId, ChunkEndorsement>>,
+}
+
+impl Client {
+    pub fn process_chunk_endorsement(
+        &mut self,
+        endorsement: ChunkEndorsement,
+    ) -> Result<(), Error> {
+        let chunk_header = self.chain.get_chunk(endorsement.chunk_hash())?.cloned_header();
+        self.chunk_endorsement_tracker.process_chunk_endorsement(chunk_header, endorsement)
+    }
+}
+
+impl EndorsementTracker {
+    pub fn new(epoch_manager: Arc<dyn EpochManagerAdapter>) -> Self {
+        Self {
+            epoch_manager,
+            chunk_endorsements: lru::LruCache::new(NUM_CHUNK_ENDORSEMENTS_CACHE_COUNT),
+        }
+    }
+
+    /// Function to process an incoming chunk endorsement from chunk validators.
+    /// We first verify the chunk endorsement and then store it in a cache.
+    /// We would later include the endorsements in the block production.
+    fn process_chunk_endorsement(
+        &mut self,
+        chunk_header: ShardChunkHeader,
+        endorsement: ChunkEndorsement,
+    ) -> Result<(), Error> {
+        let chunk_hash = endorsement.chunk_hash();
+        let account_id = &endorsement.account_id;
+
+        // If we have already processed this chunk endorsement, return early.
+        if self
+            .chunk_endorsements
+            .get(chunk_hash)
+            .is_some_and(|existing_endorsements| existing_endorsements.get(account_id).is_some())
+        {
+            tracing::debug!(target: "stateless_validation", ?endorsement, "Already received chunk endorsement.");
+            return Ok(());
+        }
+
+        if !self.epoch_manager.verify_chunk_endorsement(&chunk_header, &endorsement)? {
+            tracing::error!(target: "stateless_validation", ?endorsement, "Invalid chunk endorsement.");
+            return Err(Error::InvalidChunkEndorsement);
+        }
+
+        // If we are the current block producer, we store the chunk endorsement for each chunk which
+        // would later be used during block production to check whether to include the chunk or not.
+        // TODO(stateless_validation): It's possible for a malicious validator to send endorsements
+        // for 100 unique chunks thus pushing out current valid endorsements from our cache.
+        // Maybe add check to ensure we don't accept endorsements from chunks already included in some block?
+        // Maybe add check to ensure we don't accept endorsements from chunks that have too old height_created?
```

Comment:
Another option: deduplicate based on the (height, account_id of the chunk validator)

Reply:
Yeah that, plus a bypass for chunks that are currently being considered for inclusion in a block.

It's kind of a similar issue compared to the ShardsManager forwarding cache, we can't validate much until we wait for the header to arrive. So, one possible way to do this is to have a separate cache based on (height, account id of chunk validator) and only when we have the chunk header do we actually process those. Yeah... a bit messy.

Reply:
@staffik is working on fixing this issue. We can communicate this idea! Meanwhile for purposes of refactoring, I'm leaving this as is.